ouroboros-consensus/src/Ouroboros/Consensus/Mempool/API.hs

```diff
@@ -86,13 +86,24 @@ data Mempool m blk idx = Mempool {
       -- persistence.
       --
       -- The transactions will be validated, /in order/, before they are added
-      -- to the mempool. Invalid transactions will be returned along with the
-      -- validation error. In principle it is possible that such validation
-      -- errors are transient; for example, it is possible that a transaction is
-      -- rejected because one of its inputs is not /yet/ available in the UTxO
-      -- (the transaction it depends on is not yet in the chain, nor in the
-      -- mempool). In practice however it is likely that rejected transactions
-      -- will still be rejected later, and should just be dropped.
+      -- to the mempool. This function will return a list potentially
+      -- consisting of a few different kinds of transactions:
+      --
+      -- * Those transactions provided which were found to be valid, along
+      --   with 'Nothing' for their accompanying @Maybe (ApplyTxErr blk)@
+      --   values.
+      -- * Those transactions provided which were found to be invalid, along
+      --   with their accompanying 'ApplyTxErr's.
+      -- * Once valid transactions from the 'Mempool' which are now invalid
+      --   with respect to the ledger state, along with their accompanying
+      --   'ApplyTxErr's.
```

Comment:
BTW, the docstring still doesn't mention explicitly that (newly added) invalid transactions are *not* stored in the Mempool and that previously valid transactions that are now invalid are still in the Mempool (correct me if I'm wrong) and have to be removed. If what I'm saying is true, then that's a reason to not remove invalid transactions after `addTxs`, because the returned invalid transactions could be a mix of invalid transactions still in the mempool (to be removed) and newly invalid transaction that never made it into the mempool (no point in removing them).

Idea: have a separate type for distinguishing the three cases:
```haskell
data ValidatedTx blk
  = ValidTx     (GenTx blk)
  | InvalidTx   (GenTx blk) (ApplyTxErr blk)
  | PrevValidTx (GenTx blk) (ApplyTxErr blk)
```
(feel free to think of better names)

Reply:
I just want to make sure I understand correctly: are you saying that, within the `addTxs` implementation, we should opt to **not** remove invalid transactions (with respect to the current ledger state) which already exist within the `Mempool`?

I could see a point for this in that the return type and value is quite confusing; a mixture of the provided transactions that were valid and added to the `Mempool`, those that were invalid and not added to the `Mempool`, and also transactions that were once valid and already in the `Mempool`, but have only now been declared invalid, with respect to the current ledger state, and dropped from the `Mempool`.

I've thought about an approach similar to your suggestion (a sort of `ValidatedTx` data type) to make things more clear. However, I wasn't sure whether I should make a change such that we no longer directly return the associated type, `GenTx blk`, from `addTxs`. If we defined a `ValidatedTx` data type, `addTxs` would become something like `addTxs :: [GenTx blk] -> [ValidatedTx]` (which, by the way, I personally like a lot more).

But, then, I wonder (as you've mentioned): does it even make sense for `addTxs` to return previously known, now invalid transactions? It seems pretty counter-intuitive.

Although, if we were to go the route of not revalidating and potentially dropping existing transactions within `addTxs`, it might make the implementation a bit more complicated. We would likely have to maintain a sort of "snapshot" of the ledger state against which the `Mempool` transactions were last validated and, then, only validate the new transactions against that same snapshot (so as to guarantee that no invalid transactions, with respect to that ledger state snapshot, exist in the `Mempool`) . 

We would then have to solely rely on calls to `syncState` to ensure that the `Mempool` transactions are all valid with respect to the current ledger state of the `ChainDB`. This might not be a problem, though, since `openMempool` spawns a thread which calls `syncState` whenever the tip point of the `ChainDB` changes. However, that's an implementation detail and not defined in the API (although it is documented as of this PR).

Reply:
> I just want to make sure I understand correctly: are you saying that, within the `addTxs` implementation, we should opt to **not** remove invalid transactions (with respect to the current ledger state) which already exist within the `Mempool`?

Ah, we do intend to let `addTxs` remove invalid transactions. So I had misunderstood our intention (just like I suggested in the "correct me if I'm wrong" in my previous comment :slightly_smiling_face:). Good, so `addTxs` itself removes previously-valid-now-invalid transactions.

Then in the implementation of `addTxs`, we can trace the removed transactions (as this is done in `m` as opposed to `STM m`). If I remember correctly, being able to trace the removed transactions was the main reason of returning the previously-valid-now-invalid transactions?  Again, correct me if I'm wrong, because I might be forgetting something :slightly_smiling_face: . So if tracing was the only reason, then I don't see a reason to let `addTxs` return the previously-valid-now-invalid transactions.

The other place were previously-valid-now-invalid transactions could be removed is `syncState`, which itself cannot trace their removal (`STM m` vs `m`), so we rely on our background thread that calls it to trace them. What if "someone else" suddenly starts calling `syncState`? We would have no guarantee that the removed transactions are traced. To "fix" this, we could trace in `syncState` itself and make it live in `m` instead of `STM m`. While this scenario is unlikely, I'm not opposed to the solution, because it brings some consistency: anywhere a transaction is removed, it is being traced, and we never need to rely on the callers to do it for us. What do you think about this? Also, I think that `syncState` should live in `m` for the same reason `addTxs` lives in `m` (see its docstring, i.e. the possibility of persistence).

> I've thought about an approach similar to your suggestion (a sort of `ValidatedTx` data type) to make things more clear. However, I wasn't sure whether I should make a change such that we no longer "directly" return the associated type, `GenTx blk`, from `addTxs`. If we defined a `ValidatedTx` data type, `addTxs` would become something like `addTxs :: [GenTx blk] -> [ValidatedTx]` (which, by the way, I personally like a lot more).

It is fine to not "directly" return the associated type. We also wrap the associated type in other types like `Maybe` and `[]`, although, admittedly, they don't "hide" the associated type. Anyway, if you like `ValidatedTx` too, then go ahead and return `[ValidatedTx blk]`.

> But, then, I wonder (as you've mentioned): does it even make sense for `addTxs` to return previously known, now invalid transactions? It seems pretty counter-intuitive.
> 
> Although, if we were to go the route of not dropping existing invalid transactions within `addTxs`, it might make the implementation a bit more complicated. We would likely have to maintain a sort of "snapshot" of the ledger state against which the `Mempool` transactions were last validated and, then, only validate the new transactions against that same snapshot (so as to guarantee that no invalid transactions, with respect to that ledger state snapshot, exist in the `Mempool`) .
> 
> [..]

See my response at the top of this comment.

Reply:
> To "fix" this, we could trace in `syncState` itself and make it live in m instead of `STM m`.

But doesn't it have to live in STM? The thread that creates a new block has to get the mempool in sync with the ledger state and then grab the transactions to put into a block.

I suppose if it's specialised for that purpose then `syncState` could return the snapshot of valid txs in the mempool after syncing. Then it doesn't need to be in STM because it's atomic itself. That might well be the simplest solution.

Reply:
> > To "fix" this, we could trace in `syncState` itself and make it live in m instead of `STM m`.
> 
> But doesn't it have to live in STM? The thread that creates a new block has to get the mempool in sync with the ledger state and then grab the transactions to put into a block.

Ah, so that's the reason why it was in `STM`, I was wondering about it :slightly_smiling_face: 

> I suppose if it's specialised for that purpose then `syncState` could return the snapshot of valid txs in the mempool after syncing. Then it doesn't need to be in STM because it's atomic itself. That might well be the simplest solution.

I'm not sure I understand why this works: after obtaining the snapshot, the mempool could still change, right? Or is that not that important here?

Alternatively, we could have something like:

```haskell
withMempoolSnapshot :: (MempoolSnapshot .. -> STM m a) -> m a
```
which could be used instead of `atomically` when producing a block. It executes the given function in the same transaction as the one that syncs the mempool. After the transaction completes, the results of `syncState` are traced.